subsys/bluetooth/mesh/gen_plvl_srv.c

```diff
@@ -0,0 +1,570 @@
+	{ GEN_PLVL_OP_RANGE_SET, 4, handle_range_set },
+	{ GEN_PLVL_OP_RANGE_SET_UNACK, 4, handle_range_set_unack },
+	BT_MESH_MODEL_OP_END,
+};
+
+static void lvl_get(struct gen_lvl_srv *lvl_srv, struct bt_mesh_msg_ctx *ctx,
+		    struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+
+	srv->handlers->power_get(srv, ctx, &status);
+
+	rsp->present = POWER_TO_LVL(status.present);
+	rsp->target = POWER_TO_LVL(status.target);
+	rsp->remaining_time = status.remaining_time;
+}
+
+static void lvl_set(struct gen_lvl_srv *lvl_srv, struct bt_mesh_msg_ctx *ctx,
+		    const struct gen_lvl_set *lvl_set,
+		    struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_set set = {
+		.power = LVL_TO_POWER(lvl_set->level),
+		.transition = lvl_set->transition,
+	};
+	struct gen_plvl_status status;
+
+	change_level(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present = POWER_TO_LVL(status.present);
+		rsp->target = POWER_TO_LVL(status.target);
+		rsp->remaining_time = status.remaining_time;
+	}
+}
+
+static void lvl_delta_set(struct gen_lvl_srv *lvl_srv,
+			  struct bt_mesh_msg_ctx *ctx,
+			  const struct gen_lvl_delta_set *delta_set,
+			  struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+
+	srv->handlers->power_get(srv, NULL, &status);
+
+	struct gen_plvl_set set = {
+		.power = status.present + delta_set->delta,
+		.transition = delta_set->transition,
+	};
+
+	change_level(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present = POWER_TO_LVL(status.present);
+		rsp->target = POWER_TO_LVL(status.target);
+		rsp->remaining_time = status.remaining_time;
+	}
+}
+
+static void lvl_move_set(struct gen_lvl_srv *lvl_srv,
+			 struct bt_mesh_msg_ctx *ctx,
+			 const struct gen_lvl_move_set *move_set,
+			 struct gen_lvl_status *rsp)
+{
+	struct gen_plvl_srv *srv =
+		CONTAINER_OF(lvl_srv, struct gen_plvl_srv, lvl);
+	struct gen_plvl_status status = { 0 };
+	u16_t target;
+
+	srv->handlers->power_get(srv, NULL, &status);
+
+	if (move_set->delta > 0) {
+		target = srv->range.max;
+	} else if (move_set->delta < 0) {
+		target = srv->range.min;
+	} else {
+		target = status.present;
+	}
+
+	struct gen_plvl_set set = {
+		.power = target,
+	};
+	struct model_transition transition;
+
+	if (move_set->delta != 0 && move_set->transition) {
+		s32_t delta = target - status.present;
+
+		s32_t time_to_edge =
+			(delta * move_set->transition->time) / move_set->delta;
+
+		if (time_to_edge > 0) {
+			transition.delay = move_set->transition->delay;
+			transition.time = time_to_edge;
+			set.transition = &transition;
+		}
+	}
+
+	change_level(srv, ctx, &set, &status);
+
+	if (rsp) {
+		rsp->present = POWER_TO_LVL(status.present);
+		rsp->target = POWER_TO_LVL(status.target);
```

Comment:
This transformation will provide some value for `target` filed. This will then be passed to `gen_lvl_srv.c`, which will encode it in a wrong way, sending out incorrect status.  I suggest `target` can be copied as-is from the received message, and correct status reporting can be done in the level server itself.

Reply:
I'm not sure I follow. Is this related to the comment above, or is there some other error?

Reply:
This is some other error. Let me rephrase it a bit:

When a move status message is sent, the `target` field, if present, can have only two possible values `INT16_MAX` or `INT16_MIN`. Here, `rsp->target` value may or may not be equal to desired status values.

Reply:
Right, I see, the range might alter the target value. I think we'll need another state to get around this properly.

Reply:
May be not. As I see, as this model is _implementer_ of level server callbacks, it should provide correct values in response. Here, _gen_plvl_srv_ can compose correct `rsp->target` value, depending on the direction of the move. This will allow _gen_lvl_srv_ to send correct status.